cardano-cli/src/Cardano/CLI/Ping.hs

```diff
@@ -0,0 +1,582 @@
+import           Data.Aeson hiding (Options, json)
+import           Data.Bits (clearBit, setBit, testBit)
+import           Data.Bool (Bool (..), (&&))
+import           Data.ByteString.Lazy (ByteString)
+import           Data.Either (Either (..))
+import           Data.Eq (Eq (..))
+import           Data.Function (($), (.))
+import           Data.Functor ((<$>))
+import           Data.Int (Int)
+import           Data.List (foldl')
+import           Data.Maybe (Maybe (..), fromMaybe, isNothing)
+import           Data.Monoid (mconcat)
+import           Data.Ord (Ord (..))
+import           Data.Semigroup ((<>))
+import           Data.String (String)
+import           Data.TDigest
+import           Data.Text (Text, unpack)
+import           Data.Word (Word, Word16, Word32)
+import           GHC.Enum (Bounded (..))
+import           GHC.Err (error)
+import           GHC.Float (Double)
+import           GHC.Num (Num (..))
+import           GHC.Real (fromIntegral, realToFrac)
+import           Network.Mux.Bearer.Socket (socketAsMuxBearer)
+import           Network.Mux.Timeout (TimeoutFn, withTimeoutSerial)
+import           Network.Mux.Types
+import           Network.Socket (AddrInfo, StructLinger (..))
+import           System.Exit (ExitCode (ExitFailure))
+import           System.IO (IO)
+import           Text.Printf (printf)
+import           Text.Show (Show (..))
+
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Read as CBOR
+import qualified Codec.CBOR.Write as CBOR
+import qualified Control.Monad.Class.MonadTimer as MT
+import qualified Data.ByteString.Char8 as BS.Char
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Lazy.Char8 as BL.Char (pack, putStr)
+import qualified Data.List as L
+import qualified Network.Socket as Socket
+import qualified Options.Applicative as Opt
+import qualified Prettyprinter as PP
+import qualified System.Exit as IO
+import qualified System.IO as IO
+
+data PingCmd = PingCmd
+  { pingCmdCount    :: !Word32
+  , pingCmdHost     :: !(Maybe String)
+  , pingCmdUnixSock :: !(Maybe String)
+  , pingCmdPort     :: !String
+  , pingCmdMagic    :: !Word32
+  , pingCmdJson     :: !Bool
+  , pingCmdQuiet    :: !Bool
+  } deriving (Eq, Show)
+
+data PingCmdError = PingCmdError
+
+mainnetMagic :: Word32
+mainnetMagic = 764824073
+
+handshakeNum ::  MiniProtocolNum
+handshakeNum = MiniProtocolNum 0
+
+keepaliveNum :: MiniProtocolNum
+keepaliveNum = MiniProtocolNum 8
+
+nodeToClientVersionBit :: Int
+nodeToClientVersionBit = 15
+
+runPingCmd :: PingCmd -> ExceptT PingCmdError IO ()
+runPingCmd options = do
+  let hints = Socket.defaultHints { Socket.addrSocketType = Socket.Stream }
+
+  msgQueue <- liftIO newEmptyTMVarIO
+
+  when (isNothing (pingCmdHost options) && isNothing (pingCmdUnixSock options)) $ do
+    liftIO $ IO.putStrLn "Specify host/ip with '-h <hostname>' or a unix socket with -u <file name>"
```

Comment:
Why not propagate this error with a fleshed out `PingCmdError`?

Reply:
Done!